language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AddOnFailCodeAction.java

```diff
@@ -0,0 +1,107 @@
+import org.ballerinalang.util.diagnostic.DiagnosticErrorCode;
+import org.eclipse.lsp4j.CodeAction;
+import org.eclipse.lsp4j.Position;
+import org.eclipse.lsp4j.Range;
+import org.eclipse.lsp4j.TextEdit;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * On fail suggestion Code Action for check expressions.
+ *
+ * @since 2.0.0
+ */
+@JavaSPIService("org.ballerinalang.langserver.commons.codeaction.spi.LSCodeActionProvider")
+public class AddOnFailCodeAction extends AbstractCodeActionProvider {
+
+    public static final String NAME = "Add on-fail clause";
+
+    @Override
+    public List<CodeAction> getDiagBasedCodeActions(Diagnostic diagnostic,
+                                                    DiagBasedPositionDetails positionDetails,
+                                                    CodeActionContext context) {
+
+        if (!(DiagnosticErrorCode.CHECKED_EXPR_NO_MATCHING_ERROR_RETURN_IN_ENCL_INVOKABLE.diagnosticId()
+                .equals(diagnostic.diagnosticInfo().code())) &&
+                !(DiagnosticErrorCode.FAIL_EXPR_NO_MATCHING_ERROR_RETURN_IN_ENCL_INVOKABLE.diagnosticId()
+                        .equals(diagnostic.diagnosticInfo().code()))) {
+            return Collections.emptyList();
+        }
+
+        Node nodeAtDiagnostic = positionDetails.matchedNode();
+        List<TextEdit> edits = new ArrayList<>();
+        String commandTitle;
+
+        FailStatementResolver finder = new FailStatementResolver(diagnostic);
+        Optional<Node> failStatementResolverNode = finder.getRegComNSmtNode(nodeAtDiagnostic);
+        
+        if (failStatementResolverNode.isPresent()
+                && failStatementResolverNode.get().kind() == SyntaxKind.WHILE_KEYWORD) {
+            return Collections.emptyList();
+        }
+        
+        if (failStatementResolverNode.isPresent()) {
+            LinePosition regComStmtLinePosition = failStatementResolverNode.get().lineRange().endLine();
+            Position onFailPosLine = new Position(regComStmtLinePosition.line(), regComStmtLinePosition.offset());
+
+            String spaces = " ".repeat(regComStmtLinePosition.offset() - 1);
+            String editText = " on fail var e {" + CommonUtil.LINE_SEPARATOR + spaces + "\t"
+                    + CommonUtil.LINE_SEPARATOR + spaces + "}";
+            edits.add(new TextEdit(new Range(onFailPosLine, onFailPosLine), editText));
+            commandTitle = CommandConstants.CREATE_ON_FAIL_CLAUSE;
+        } else {
+            while (!StatementNode.class.isAssignableFrom(nodeAtDiagnostic.getClass())) {
+                nodeAtDiagnostic = nodeAtDiagnostic.parent();
+            }
+            LinePosition linePosition = nodeAtDiagnostic.lineRange().startLine();
+            Position positionDo = new Position(linePosition.line(), linePosition.offset());
+            Position posCheckLineStart = new Position(linePosition.line() + 1, 0);
```

Comment:
We can't always say that a statement node is single line. Instead we should do ,
```
Position positionDo = new Position(nodeAtDiagnostic.lineRange().startLine().line(), nodeAtDiagnostic.lineRange().startLine().offset());
Position posCheckLineStart = new Position(nodeAtDiagnostic.lineRange().endLine().line(), nodeAtDiagnostic.lineRange().endLine().offset());
```